src_blog/_posts/Grain-Formatter.md

```diff
@@ -0,0 +1,120 @@
+
+## Implementation
+
+The formatter works by using the Grain compiler to first parse the source code into what's called an Abstract Syntax Tree, or AST, which captures the meaning of the program as a tree of patterns, statements, and expressions. Then we just write the tree back out as valid Grain code. That's sometimes easier said than done, and here are some of the challenges we faced:
+
+
+### Challenges
+
+-	Comments.   Comments are, by definition, ignored when the compiler is comprehending the meaning of our code, so they aren't stored in the AST. When formatting, however, we need to preserve them. To this end, we extract these comments and combine them with the AST and bring the regenerated code and comments back together, checking how the code looks along the way. In some of the more esoteric applications of comments, their placement can be ambiguous in relation to source code that also doesn’t appear in the AST, for example:
+
+```sh
+variable1 /* comment */ , variable2
+```
+
+Our AST encodes the locations of the variable names and the comment, but we don’t know if the comment comes before or after the comma.
+
+-	Grain uses various forms of syntactic sugar.  One of my favourites is converting list syntax of `[1,2,3]` into repeated application of the list `cons` operator. The formatter needs to be aware of these and rewrite the AST back out in sugared syntax.
+
+- Blank lines are useful for structuring and organising code but also need to be styled for consistency.  Removing all blank lines really impacts on how the source code looks and reads. We chose to compress multiple blank lines down into a single blank line, but to always retain at least one blank line wherever one was found.
+
+-	Arguments over style. As you might imagine, we had quite a lot of discussion about style choices. There really is no single way that suits everyone, and whatever is chosen will always be a compromise. However, what we wanted to avoid was adding configurable formatting where different formatting choices can be specified.  This really undermines the premise of all formatted Grain code looking the same, which we think really helps all Grain programmers, new or experienced, to look at Grain code and find it easy to read through because of the consistency, be than in end user code or the standard libraries.
+-	How far should the rewriting go? Some formatters go further than we have in rewriting code. For example, we’ve seen some which will examine `if` statements and rewrite them using ternary operators if appropriate. So far, we’ve not gone that far. There are a couple of places where we might make changes from the original code. For example, syntactic sugar for statements such as `x += 1` may have been written as `x = x + 1`; both are identical in the AST representation, and without examining the unparsed source we can’t tell what the coder actually wrote. Feedback on whether this is acceptable, good, or needs fixing will be very welcome!
```

Comment:
> rewrite them using ternary operators if appropriate. So far, we’ve not gone that far.

Should we say this? Grain doesn't currently have (or really even have plans to support) a ternary operator.

Reply:
We can cut it, it's a example of what I remember the Reason reformatter used to do.

Reply:
I changed to to make clearer this was for other languages.